thesis/03-implementation.tex

```diff
@@ -83,23 +83,491 @@ \subsubsection{Backend}
+\begin{itemize}
+	\item authentication
+	\item account creation
+	\item password resetting 
+	\item password changing
+\end{itemize}
+
+To fulfill its duties, the Accounts sub-application exposes several REST API endpoints.
+The data model is mostly derived from the built-in Django model increasing the development speed considerably.
+\medskip
+
+ 
+\begin{figure}[H]
+  \caption{Accounts - ERD diagram}
+  \centering
+    \includegraphics[width=\textwidth]{assets/3-erd-accounts.png}
+\end{figure}
+
+
+
+
 \subsection{Algorithms}
 % Ciekawsze algorytmy, aspekty, mechanizmy np. logowanie, indeksowane, cachowanie, synchronizacja, backup, jakieś procesy w tle, jakieś progress bary, jakieś analizy, generowanie warstw GISowych, lokalizacja etc. Jest tu często o czym pisać. %
 
-\subsection{Development environment setup}
+\subsection{Quality assurance}
+
+To ensure the quality of the application, the codebase is continuously linted, tested, built, and delivered. Immediate feedback is available for developers, code reviewers, and operators, reducing the possibility of merging a faulty code into the \texttt{master} branch and therefore using a faulty deployment in the production environment.
+
+\medskip
+The pipeline is implemented using \tech{Github Actions}. \tech{Github Actions} comes bundled with \tech{Github}, allowing for seamless integration with the code repository.
+
+\medskip
+Furthermore, no changes can be pushed directly to the \texttt{master} branch. Every commit needs to be reviewed by at least one other developer before being merged. This policy is enforced using \tech{Github} settings.
+
+\medskip
+Each system component has a dedicated pipeline, tailored to specific requirements. 
+
+\medskip
+Continuous delivery and deployment reduce the risk of deploying an application that has not passed all the required checks in a pipeline.
+
+\medskip
+All of the dependencies are bound to a specific minor version. Doing so ensures that in case one of the vendors implements a breaking change our application and pipelines do not break.
+
+\subsubsection{Frontend}
+
+The frontend code is linted on every commit using \tech{Husky}, \tech{ESLint}, and \tech{Prettier}. It is not possible to push an incorrectly formatted code to the repository.
+Uniformly formatted code improves the clarity, allowing a code reviewer to focus on business logic and potential bugs instead of code-formatting.
+
+\medskip
+Most of the business logic on the frontend is tested. The test runner of choice is \tech{Jest}.
+
+\medskip
+
+To aid the code review process, every pull request with changes related to the frontend component is deployed as a \tech{preview application} on \tech{Netlify}. A \tech{preview application} is an isolated environment with an application built using the code submitted along with a pull request.
+
+\medskip
+
+Running the whole test suite is as simple as executing \texttt{yarn test}.
```

Comment:
I'd rearrange the paragraphs, to have a more logical connection.

First, "the frontend code is linted...", then "To aid the code review" (since you ended the first talking about a reviewer), then "Most of the business logic...", and connect it with "running the whole test..."

Reply:
Fixed. Still sounds strange tho.